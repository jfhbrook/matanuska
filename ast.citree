import { Token } from 'typescript-parsec'
import { TokenKind } from './scanner'

type Expr in './expr' {
  LineNo         => value: number
  DecimalLiteral => value: number
  HexLiteral     => value: number
  OctalLiteral   => value: number
  BinaryLiteral  => value: number
  RealLiteral    => value: number
  BooleanLiteral => value: boolean
  StringLiteral  => value: string
}

type Stmt in './stmt' {
  import { Expr, LineNo } from './expr'

  Line => lineNo: LineNo, command: Expr
  Print => expression: Expr
}
