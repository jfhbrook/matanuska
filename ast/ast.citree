import { Token, TokenKind } from '../tokens'

type Expr in './expr' {
  IntLiteral     => value: number
  RealLiteral    => value: number
  BoolLiteral    => value: boolean
  StringLiteral  => value: string
  NilLiteral!
}

type Cmd in './cmd' {
  import { Expr } from './expr'

  Expression => expression: Expr
  Print      => expression: Expr
}
