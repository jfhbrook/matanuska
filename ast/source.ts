//
// A tuple-like class to store source code. This is needed so we can easily
// renumber source lines without having to use regular expressions.
//

export class Source {
  constructor(
    public leadingWs: string,
    public lineNo: string,
    public separatingWs: string,
    public source: string,
  ) {}

  public clone(): Source {
    return new Source(
      this.leadingWs,
      this.lineNo,
      this.separatingWs,
      this.source,
    );
  }

  public toString(): string {
    return [this.leadingWs, this.lineNo, this.separatingWs, this.source].join(
      '',
    );
  }
}
