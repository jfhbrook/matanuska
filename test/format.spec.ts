import * as assert from 'assert';

import { describe, test } from 'vitest';
import { t } from './helpers/tap';

import {
  Formatter,
  DefaultFormatter,
  Inspector,
  inspectString,
} from '../format';

import { ErrorCode } from '../errors';
import {
  BaseException,
  Exception,
  RuntimeError,
  TypeError,
  NotImplementedError,
  AssertionError,
  BaseWarning,
  ZeroDivisionError,
  OsError,
  FileError,
  SyntaxError,
  ParseError,
  SyntaxWarning,
  ParseWarning,
} from '../exceptions';
import { Exit, ExitCode } from '../exit';
import { BaseFault, RuntimeFault, UsageFault } from '../faults';
import { Token, TokenKind } from '../tokens';
import { Source } from '../ast/source';
import {
  Binary,
  Unary,
  Variable,
  IntLiteral,
  RealLiteral,
  BoolLiteral,
  StringLiteral,
  PromptLiteral,
  NilLiteral,
} from '../ast/expr';
import {
  Print,
  Exit as ExitInstr,
  Rem,
  New,
  Load,
  List,
  Save,
  Run,
  Let,
  Assign,
  ShortIf,
  If,
  Else,
  ElseIf,
  EndIf,
  For,
  Next,
  EndFor,
  While,
  EndWhile,
  Repeat,
  Until,
  End,
} from '../ast/instr';
import { Line, Program } from '../ast';
import { FILENAME } from './helpers/files';
import { scrubNodeVersion } from './helpers/format';
import { TRACEBACK } from './helpers/traceback';

const LINE = '100 print someFn(ident';
const IS_LINE = [true, false];

describe('inspectString', () => {
  test('it inspects strings without quotes', () => {
    t.equal(inspectString('hello'), "'hello'");
  });
  test('it inspects strings with single quotes', () => {
    t.equal(inspectString("don't"), '"don\'t"');
  });

  test('it inspects strings with double quotes', () => {
    t.equal(inspectString('"time machine"'), '\'"time machine"\'');
  });

  test('it inspects strings with both kinds of quotes', () => {
    t.equal(
      inspectString('don\'t mess with my "time machine"'),
      "'don\\'t mess with my \"time machine\"'",
    );
  });
});

function formatTestSuite<F extends Formatter>(formatter: F): void {
  describe(`given a ${formatter.constructor.name}`, () => {
    test('it formats a string', () => {
      t.matchSnapshot(formatter.format('hello'));
    });

    test('it formats a number', () => {
      t.matchSnapshot(formatter.format(12345));
    });

    test('it formats a boolean', () => {
      t.matchSnapshot(formatter.format(true));
    });

    test('it formats a Traceback', () => {
      t.matchSnapshot(formatter.format(TRACEBACK));
    });

    for (const ctor of [
      BaseException,
      Exception,
      RuntimeError,
      NotImplementedError,
    ]) {
      test(`it formats a ${ctor.prototype.name}`, () => {
        t.matchSnapshot(formatter.format(new ctor('message', TRACEBACK)));
      });
    }

    test('it formats a BaseWarning', () => {
      t.matchSnapshot(formatter.format(new BaseWarning('message', TRACEBACK)));
    });

    test('it formats a BaseWarning without a traceback', () => {
      t.matchSnapshot(formatter.format(new BaseWarning('message', null)));
    });

    test('it formats an AssertionError', () => {
      t.matchSnapshot(
        formatter.format(new AssertionError('message', TRACEBACK)),
      );
    });

    test('it formats a TypeError', () => {
      t.matchSnapshot(
        formatter.format(
          new TypeError('message', 123, 'integer', 'nil', TRACEBACK),
        ),
      );
    });

    test('it formats a ZeroDivisionError', () => {
      t.matchSnapshot(
        formatter.format(
          new ZeroDivisionError(1, 'integer', 0, 'integer', TRACEBACK),
        ),
      );
    });

    test('it formats an OsError', () => {
      t.matchSnapshot(
        formatter.format(
          new OsError('message', ErrorCode.AddressInUse, null, TRACEBACK),
        ),
      );
    });

    test('it formats a FileError with one file', () => {
      t.matchSnapshot(
        formatter.format(
          new FileError(
            'message',
            ErrorCode.Access,
            ExitCode.NoInput,
            [FILENAME],
            TRACEBACK,
          ),
        ),
      );
    });

    test('it formats a FileError with two files', () => {
      t.matchSnapshot(
        formatter.format(
          new FileError(
            'message',
            ErrorCode.Access,
            ExitCode.NoInput,
            [FILENAME, 'another.bas'],
            TRACEBACK,
          ),
        ),
      );
    });

    for (const isLine of IS_LINE) {
      describe(`when it ${isLine ? 'is' : 'is not'} a line`, () => {
        const line = isLine
          ? Source.command(LINE)
          : Source.command(LINE.replace(/^\d+ /, ''));
        test('it formats a SyntaxError', () => {
          t.matchSnapshot(
            formatter.format(
              new SyntaxError('expected )', {
                filename: FILENAME,
                row: 0,
                isLine,
                lineNo: 100,
                cmdNo: isLine ? 10 : null,
                offsetStart: 22,
                offsetEnd: 23,
                source: line,
              }),
            ),
          );
        });

        test('it formats a ParseError', () => {
          t.matchSnapshot(
            formatter.format(
              new ParseError([
                new SyntaxError('expected )', {
                  filename: FILENAME,
                  row: 0,
                  isLine,
                  lineNo: 100,
                  cmdNo: isLine ? 10 : null,
                  offsetStart: 22,
                  offsetEnd: 23,
                  source: line,
                }),
                new SyntaxWarning('identifier has no sigil', {
                  filename: FILENAME,
                  row: 0,
                  isLine,
                  lineNo: 100,
                  cmdNo: isLine ? 10 : null,
                  offsetStart: 17,
                  offsetEnd: 18,
                  source: line,
                }),
              ]),
            ),
          );
        });

        test('it formats a SyntaxWarning', () => {
          t.matchSnapshot(
            formatter.format(
              new SyntaxWarning('expected )', {
                filename: FILENAME,
                row: 0,
                isLine,
                lineNo: 100,
                cmdNo: isLine ? 10 : null,
                offsetStart: 22,
                offsetEnd: 23,
                source: line,
              }),
            ),
          );
        });

        test('it formats a ParseWarning', () => {
          t.matchSnapshot(
            formatter.format(
              new ParseWarning([
                new SyntaxWarning('identifier has no sigil', {
                  filename: FILENAME,
                  row: 0,
                  isLine,
                  lineNo: 100,
                  cmdNo: isLine ? 10 : null,
                  offsetStart: 17,
                  offsetEnd: 18,
                  source: line,
                }),
              ]),
            ),
          );
        });
      });
    }

    test('it formats a BaseFault', () => {
      t.matchSnapshot(formatter.format(new BaseFault('message', TRACEBACK)));
    });

    test('it formats a RuntimeFault', () => {
      const underlying = new assert.AssertionError({
        message: 'underlying assertion',
        actual: false,
        expected: true,
        operator: '===',
      });

      t.matchSnapshot(
        scrubNodeVersion(
          formatter.format(
            new RuntimeFault('Some runtime fault', underlying, TRACEBACK),
          ),
        ),
      );
    });

    test('it formats a UsageFault', () => {
      t.matchSnapshot(formatter.format(new UsageFault('Usage: lol')));
    });

    test('it formats an Exit with a message', () => {
      t.matchSnapshot(formatter.format(new Exit(ExitCode.Success, 'message')));
    });

    test('it formats an Exit without message', () => {
      t.matchSnapshot(formatter.format(new Exit(ExitCode.Success)));
    });

    test('it formats a Token', () => {
      t.matchSnapshot(
        formatter.format(
          new Token({
            kind: TokenKind.StringLiteral,
            index: 0,
            row: 0,
            offsetStart: 0,
            offsetEnd: 9,
            text: "'hello\\q'",
            value: 'hello\\q',
          }),
        ),
      );
    });

    test('it formats a Unary expression', () => {
      t.matchSnapshot(
        formatter.format(new Unary(TokenKind.Minus, new IntLiteral(1))),
      );
    });

    test('it formats a Binary expression', () => {
      t.matchSnapshot(
        formatter.format(
          new Binary(new IntLiteral(1), TokenKind.Plus, new IntLiteral(1)),
        ),
      );
    });

    test.todo('it formats a Logical expression', () => {
      /*
      t.matchSnapshot(formatter.format(new Logical(
        new IntLiteral(1),
        TokenKind.And,
        new IntLiteral(1)
      )));
      */
    });

    test('it formats an IntLiteral', () => {
      t.matchSnapshot(formatter.format(new IntLiteral(12)));
    });

    test('it formats a RealLiteral', () => {
      t.matchSnapshot(formatter.format(new RealLiteral(123.456)));
    });

    test('it formats a BoolLiteral', () => {
      t.matchSnapshot(formatter.format(new BoolLiteral(true)));
    });

    test('it formats a StringLiteral', () => {
      t.matchSnapshot(formatter.format(new StringLiteral('hello')));
    });

    test('it formats a PromptLiteral', () => {
      t.matchSnapshot(formatter.format(new PromptLiteral('\\u@\\h:\\w\\$')));
    });

    test('it formats a NilLiteral', () => {
      t.matchSnapshot(formatter.format(new NilLiteral()));
    });

    test('it formats a Instr', () => {
      t.matchSnapshot(formatter.format(new Print(new StringLiteral('hello'))));
    });

    test('it formats a Line', () => {
      t.matchSnapshot(
        formatter.format(
          new Line(100, 1, new Source('', '100', ' ', 'print "hello world"'), [
            new Print(new StringLiteral('hello world')),
          ]),
        ),
      );
    });

    test('it formats a Program', () => {
      t.matchSnapshot(
        formatter.format(
          new Program(FILENAME, [
            new Line(
              100,
              1,
              new Source('', '100', ' ', 'print "hello world"'),
              [new Print(new StringLiteral('hello world'))],
            ),
          ]),
        ),
      );
    });

    test('instructions', () => {
      const INSTRUCTIONS = [
        new Print(new StringLiteral('hello')),
        new ExitInstr(new IntLiteral(0)),
        new Rem('a witty remark'),
        new New(null),
        new Load(new StringLiteral('./examples/001-hello-world.bas'), true),
        new List(),
        new Save(null),
        new Run(),
        new Let(
          new Variable(
            new Token({
              kind: TokenKind.IntIdent,
              index: 0,
              row: 1,
              offsetStart: 5,
              offsetEnd: 6,
              text: 'i%',
              value: null,
            }),
          ),
          new IntLiteral(1),
        ),
        new Assign(
          new Variable(
            new Token({
              kind: TokenKind.IntIdent,
              index: 0,
              row: 1,
              offsetStart: 0,
              offsetEnd: 1,
              text: 'i%',
              value: null,
            }),
          ),
          new IntLiteral(1),
        ),
        new ShortIf(
          new BoolLiteral(true),
          [new Print(new StringLiteral('true'))],
          [],
        ),
        new ShortIf(
          new BoolLiteral(true),
          [new Print(new StringLiteral('true'))],
          [new Print(new StringLiteral('false'))],
        ),
        new If(new BoolLiteral(true)),
        new Else(),
        new ElseIf(new BoolLiteral(true)),
        new EndIf(),
        new For(
          new Variable(
            new Token({
              kind: TokenKind.IntIdent,
              index: 0,
              row: 1,
              offsetStart: 5,
              offsetEnd: 6,
              text: 'i%',
              value: null,
            }),
          ),
          new IntLiteral(1),
        ),
        new Next(),
        new EndFor(),
        new While(new BoolLiteral(true)),
        new EndWhile(),
        new Repeat(),
        new Until(new BoolLiteral(false)),
        new End(),
      ];

      for (const instr of INSTRUCTIONS) {
        t.matchSnapshot(formatter.format(instr));
      }
    });

    test('it formats a native value', () => {
      t.matchSnapshot(formatter.format(new Set('abc')));
    });

    test('it formats a null value', () => {
      t.matchSnapshot(formatter.format(null));
    });

    test('it formats an undefined value', () => {
      t.matchSnapshot(formatter.format(undefined));
    });

    test('it formats an array of values', () => {
      t.matchSnapshot(formatter.format([1, 'two', true]));
    });
  });
}

for (const ctor of [DefaultFormatter, Inspector]) {
  formatTestSuite(new ctor());
}
