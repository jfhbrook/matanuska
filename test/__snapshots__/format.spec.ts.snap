// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given a DefaultFormatter > instructions 1`] = `"Print('hello')"`;

exports[`given a DefaultFormatter > instructions 2`] = `"Exit(0)"`;

exports[`given a DefaultFormatter > instructions 3`] = `"Rem(a witty remark)"`;

exports[`given a DefaultFormatter > instructions 4`] = `"New(null)"`;

exports[`given a DefaultFormatter > instructions 5`] = `"Load('./examples/001-hello-world.bas', run=true)"`;

exports[`given a DefaultFormatter > instructions 6`] = `"List"`;

exports[`given a DefaultFormatter > instructions 7`] = `"New(null)"`;

exports[`given a DefaultFormatter > instructions 8`] = `"Run"`;

exports[`given a DefaultFormatter > instructions 9`] = `"Let(i%, 1)"`;

exports[`given a DefaultFormatter > instructions 10`] = `"ShortIf (true) { Print('true') }"`;

exports[`given a DefaultFormatter > instructions 11`] = `"ShortIf (true) { Print('true') } { Print('false') }"`;

exports[`given a DefaultFormatter > instructions 12`] = `"If (true)"`;

exports[`given a DefaultFormatter > instructions 13`] = `"Else"`;

exports[`given a DefaultFormatter > instructions 14`] = `"ElseIf (true)"`;

exports[`given a DefaultFormatter > instructions 15`] = `"End"`;

exports[`given a DefaultFormatter > it formats a BaseException 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
BaseException: message"
`;

exports[`given a DefaultFormatter > it formats a BaseFault 1`] = `
"=== INTERNAL FAULT ===

--- Internal Stack Trace ---
BaseFault: message
    at {CWD}/test/format.spec.ts:277:10
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:146:14
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:533:11
    at runWithTimeout (file://{CWD}/node_modules/@vitest/runner/dist/index.js:61:7)
    at runTest (file://{CWD}/node_modules/@vitest/runner/dist/index.js:986:17)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runFiles (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1199:5)
    at startTests (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1208:3)
    at file://{CWD}/node_modules/vitest/dist/chunks/runBaseTests.D-Gcin7G.js:130:11"
`;

exports[`given a DefaultFormatter > it formats a BaseWarning 1`] = `"/home/josh/script.bas:100: BaseWarning: message"`;

exports[`given a DefaultFormatter > it formats a BaseWarning without a traceback 1`] = `"<unknown>:<?>: BaseWarning: message"`;

exports[`given a DefaultFormatter > it formats a Binary expression 1`] = `
"Binary(+) {
  1,
  1,
}"
`;

exports[`given a DefaultFormatter > it formats a BoolLiteral 1`] = `"true"`;

exports[`given a DefaultFormatter > it formats a Exception 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
Exception: message"
`;

exports[`given a DefaultFormatter > it formats a FileError with one file 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
FileError EACCES: message
  Input File: /home/josh/script.bas"
`;

exports[`given a DefaultFormatter > it formats a FileError with two files 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
FileError EACCES: message
  Source File: /home/josh/script.bas
  Destination File: another.bas"
`;

exports[`given a DefaultFormatter > it formats a Instr 1`] = `"Print('hello')"`;

exports[`given a DefaultFormatter > it formats a Line 1`] = `
"Line(100) [
  Print('hello world'),
]"
`;

exports[`given a DefaultFormatter > it formats a NilLiteral 1`] = `"nil"`;

exports[`given a DefaultFormatter > it formats a NotImplementedError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
NotImplementedError: message"
`;

exports[`given a DefaultFormatter > it formats a Program 1`] = `
"Program(
  Line(100) [
    Print('hello world'),
  ],
)"
`;

exports[`given a DefaultFormatter > it formats a PromptLiteral 1`] = `"'\\u@\\h:\\w\\$'"`;

exports[`given a DefaultFormatter > it formats a RealLiteral 1`] = `"123.456"`;

exports[`given a DefaultFormatter > it formats a RuntimeError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
RuntimeError: message"
`;

exports[`given a DefaultFormatter > it formats a RuntimeFault 1`] = `
"=== RUNTIME FAULT ===

--- Internal Stack Trace ---
AssertionError [ERR_ASSERTION]: underlying assertion
    at new AssertionError (node:internal/assert/assertion_error:447:5)
    at {CWD}/test/format.spec.ts:279:18
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:146:14
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:533:11
    at runWithTimeout (file://{CWD}/node_modules/@vitest/runner/dist/index.js:61:7)
    at runTest (file://{CWD}/node_modules/@vitest/runner/dist/index.js:986:17)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runFiles (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1199:5)
    at startTests (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1208:3)

--- Traceback ---
Traceback:
  File '/home/josh/script.bas', line 100

--- Versions ---
Matanuska BASIC: v0.0.1
TypeScript: v5.5.2
Node.js: NODE_VERSION

This is a bug in Matanuska BASIC. If you copy this entire message and post it
to the issues tracker:

    https://github.com/jfhbrook/matanuska/issues

the developers will do their best to fix it - and more importantly, they will
owe you a beer, coffee or beverage of your choice. 🍻

"
`;

exports[`given a DefaultFormatter > it formats a StringLiteral 1`] = `"'hello'"`;

exports[`given a DefaultFormatter > it formats a Token 1`] = `
"Token(<string>) {
  index: 0,
  row: 0,
  offsetStart: 0,
  offsetEnd: 9,
  text: "'hello\\q'",
  value: 'hello\\q',
}"
`;

exports[`given a DefaultFormatter > it formats a Traceback 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100"
`;

exports[`given a DefaultFormatter > it formats a TypeError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
TypeError: message
  Value: 123
  From: integer
  To: nil"
`;

exports[`given a DefaultFormatter > it formats a Unary expression 1`] = `
"Unary(-) {
  1,
}"
`;

exports[`given a DefaultFormatter > it formats a UsageFault 1`] = `"Usage: lol"`;

exports[`given a DefaultFormatter > it formats a ZeroDivisionError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
ZeroDivisionError: Cannot divide 1 by 0"
`;

exports[`given a DefaultFormatter > it formats a boolean 1`] = `"true"`;

exports[`given a DefaultFormatter > it formats a native value 1`] = `"Set(3) { 'a', 'b', 'c' }"`;

exports[`given a DefaultFormatter > it formats a null value 1`] = `"null"`;

exports[`given a DefaultFormatter > it formats a number 1`] = `"12345"`;

exports[`given a DefaultFormatter > it formats a string 1`] = `"hello"`;

exports[`given a DefaultFormatter > it formats an AssertionError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
AssertionError: message"
`;

exports[`given a DefaultFormatter > it formats an Exit with a message 1`] = `"Exit 0: message"`;

exports[`given a DefaultFormatter > it formats an Exit without message 1`] = `"Exit 0"`;

exports[`given a DefaultFormatter > it formats an IntLiteral 1`] = `"12"`;

exports[`given a DefaultFormatter > it formats an OsError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
OsError EADDRINUSE: message"
`;

exports[`given a DefaultFormatter > it formats an array of values 1`] = `
"[
  1,
  'two',
  true,
]"
`;

exports[`given a DefaultFormatter > it formats an undefined value 1`] = `"undefined"`;

exports[`given a DefaultFormatter > when it is a line > it formats a ParseError 1`] = `
"/home/josh/script.bas:100:22: error: expected )
  100 print someFn(ident
                        ^
/home/josh/script.bas:100:17: warning: identifier has no sigil
  100 print someFn(ident
                   ^"
`;

exports[`given a DefaultFormatter > when it is a line > it formats a ParseWarning 1`] = `
"/home/josh/script.bas:100:17: warning: identifier has no sigil
  100 print someFn(ident
                   ^"
`;

exports[`given a DefaultFormatter > when it is a line > it formats a SyntaxError 1`] = `
"/home/josh/script.bas:100:22: error: expected )
  100 print someFn(ident
                        ^"
`;

exports[`given a DefaultFormatter > when it is a line > it formats a SyntaxWarning 1`] = `
"/home/josh/script.bas:100:22: warning: expected )
  100 print someFn(ident
                        ^"
`;

exports[`given a DefaultFormatter > when it is not a line > it formats a ParseError 1`] = `
"/home/josh/script.bas:<R0>:22 (after line 100): error: expected )
  print someFn(ident
                        ^
/home/josh/script.bas:<R0>:17 (after line 100): warning: identifier has no sigil
  print someFn(ident
                   ^"
`;

exports[`given a DefaultFormatter > when it is not a line > it formats a ParseWarning 1`] = `
"/home/josh/script.bas:<R0>:17 (after line 100): warning: identifier has no sigil
  print someFn(ident
                   ^"
`;

exports[`given a DefaultFormatter > when it is not a line > it formats a SyntaxError 1`] = `
"/home/josh/script.bas:<R0>:22 (after line 100): error: expected )
  print someFn(ident
                        ^"
`;

exports[`given a DefaultFormatter > when it is not a line > it formats a SyntaxWarning 1`] = `
"/home/josh/script.bas:<R0>:22 (after line 100): warning: expected )
  print someFn(ident
                        ^"
`;

exports[`given a Inspector > instructions 1`] = `"Print('hello')"`;

exports[`given a Inspector > instructions 2`] = `"Exit(0)"`;

exports[`given a Inspector > instructions 3`] = `"Rem(a witty remark)"`;

exports[`given a Inspector > instructions 4`] = `"New([1mnull[22m)"`;

exports[`given a Inspector > instructions 5`] = `"Load('./examples/001-hello-world.bas', run=true)"`;

exports[`given a Inspector > instructions 6`] = `"List"`;

exports[`given a Inspector > instructions 7`] = `"New([1mnull[22m)"`;

exports[`given a Inspector > instructions 8`] = `"Run"`;

exports[`given a Inspector > instructions 9`] = `"Let(i%, 1)"`;

exports[`given a Inspector > instructions 10`] = `"ShortIf (true) { Print('true') }"`;

exports[`given a Inspector > instructions 11`] = `"ShortIf (true) { Print('true') } { Print('false') }"`;

exports[`given a Inspector > instructions 12`] = `"If (true)"`;

exports[`given a Inspector > instructions 13`] = `"Else"`;

exports[`given a Inspector > instructions 14`] = `"ElseIf (true)"`;

exports[`given a Inspector > instructions 15`] = `"End"`;

exports[`given a Inspector > it formats a BaseException 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
BaseException: message"
`;

exports[`given a Inspector > it formats a BaseFault 1`] = `
"=== INTERNAL FAULT ===

--- Internal Stack Trace ---
BaseFault: message
    at {CWD}/test/format.spec.ts:277:10
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:146:14
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:533:11
    at runWithTimeout (file://{CWD}/node_modules/@vitest/runner/dist/index.js:61:7)
    at runTest (file://{CWD}/node_modules/@vitest/runner/dist/index.js:986:17)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runFiles (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1199:5)
    at startTests (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1208:3)
    at file://{CWD}/node_modules/vitest/dist/chunks/runBaseTests.D-Gcin7G.js:130:11"
`;

exports[`given a Inspector > it formats a BaseWarning 1`] = `"/home/josh/script.bas:100: BaseWarning: message"`;

exports[`given a Inspector > it formats a BaseWarning without a traceback 1`] = `"<unknown>:<?>: BaseWarning: message"`;

exports[`given a Inspector > it formats a Binary expression 1`] = `
"Binary(+) {
  1,
  1,
}"
`;

exports[`given a Inspector > it formats a BoolLiteral 1`] = `"true"`;

exports[`given a Inspector > it formats a Exception 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
Exception: message"
`;

exports[`given a Inspector > it formats a FileError with one file 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
FileError EACCES: message
  Input File: /home/josh/script.bas"
`;

exports[`given a Inspector > it formats a FileError with two files 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
FileError EACCES: message
  Source File: /home/josh/script.bas
  Destination File: another.bas"
`;

exports[`given a Inspector > it formats a Instr 1`] = `"Print('hello')"`;

exports[`given a Inspector > it formats a Line 1`] = `
"Line(100) [
  Print('hello world'),
]"
`;

exports[`given a Inspector > it formats a NilLiteral 1`] = `"nil"`;

exports[`given a Inspector > it formats a NotImplementedError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
NotImplementedError: message"
`;

exports[`given a Inspector > it formats a Program 1`] = `
"Program(
  Line(100) [
    Print('hello world'),
  ],
)"
`;

exports[`given a Inspector > it formats a PromptLiteral 1`] = `"'\\u@\\h:\\w\\$'"`;

exports[`given a Inspector > it formats a RealLiteral 1`] = `"123.456"`;

exports[`given a Inspector > it formats a RuntimeError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
RuntimeError: message"
`;

exports[`given a Inspector > it formats a RuntimeFault 1`] = `
"=== RUNTIME FAULT ===

--- Internal Stack Trace ---
AssertionError [ERR_ASSERTION]: underlying assertion
    at new AssertionError (node:internal/assert/assertion_error:447:5)
    at {CWD}/test/format.spec.ts:279:18
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:146:14
    at file://{CWD}/node_modules/@vitest/runner/dist/index.js:533:11
    at runWithTimeout (file://{CWD}/node_modules/@vitest/runner/dist/index.js:61:7)
    at runTest (file://{CWD}/node_modules/@vitest/runner/dist/index.js:986:17)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runSuite (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1142:15)
    at runFiles (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1199:5)
    at startTests (file://{CWD}/node_modules/@vitest/runner/dist/index.js:1208:3)

--- Traceback ---
Traceback:
  File '/home/josh/script.bas', line 100

--- Versions ---
Matanuska BASIC: v0.0.1
TypeScript: v5.5.2
Node.js: NODE_VERSION

This is a bug in Matanuska BASIC. If you copy this entire message and post it
to the issues tracker:

    https://github.com/jfhbrook/matanuska/issues

the developers will do their best to fix it - and more importantly, they will
owe you a beer, coffee or beverage of your choice. 🍻

"
`;

exports[`given a Inspector > it formats a StringLiteral 1`] = `"'hello'"`;

exports[`given a Inspector > it formats a Token 1`] = `
"Token(<string>) {
  index: 0,
  row: 0,
  offsetStart: 0,
  offsetEnd: 9,
  text: "'hello\\q'",
  value: 'hello\\q',
}"
`;

exports[`given a Inspector > it formats a Traceback 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100"
`;

exports[`given a Inspector > it formats a TypeError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
TypeError: message
  Value: [33m123[39m
  From: integer
  To: nil"
`;

exports[`given a Inspector > it formats a Unary expression 1`] = `
"Unary(-) {
  1,
}"
`;

exports[`given a Inspector > it formats a UsageFault 1`] = `"Usage: lol"`;

exports[`given a Inspector > it formats a ZeroDivisionError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
ZeroDivisionError: Cannot divide 1 by 0"
`;

exports[`given a Inspector > it formats a boolean 1`] = `"[33mtrue[39m"`;

exports[`given a Inspector > it formats a native value 1`] = `"Set(3) { [32m'a'[39m, [32m'b'[39m, [32m'c'[39m }"`;

exports[`given a Inspector > it formats a null value 1`] = `"[1mnull[22m"`;

exports[`given a Inspector > it formats a number 1`] = `"[33m12345[39m"`;

exports[`given a Inspector > it formats a string 1`] = `"[32m'hello'[39m"`;

exports[`given a Inspector > it formats an AssertionError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
AssertionError: message"
`;

exports[`given a Inspector > it formats an Exit with a message 1`] = `"Exit [32m0[39m: message"`;

exports[`given a Inspector > it formats an Exit without message 1`] = `"Exit [32m0[39m"`;

exports[`given a Inspector > it formats an IntLiteral 1`] = `"12"`;

exports[`given a Inspector > it formats an OsError 1`] = `
"Traceback:
  File '/home/josh/script.bas', line 100
OsError EADDRINUSE: message"
`;

exports[`given a Inspector > it formats an array of values 1`] = `
"[
  [33m1[39m,
  [32m'two'[39m,
  [33mtrue[39m,
]"
`;

exports[`given a Inspector > it formats an undefined value 1`] = `"[90mundefined[39m"`;

exports[`given a Inspector > when it is a line > it formats a ParseError 1`] = `
"/home/josh/script.bas:100:22: error: expected )
  100 print someFn(ident
                        ^
/home/josh/script.bas:100:17: warning: identifier has no sigil
  100 print someFn(ident
                   ^"
`;

exports[`given a Inspector > when it is a line > it formats a ParseWarning 1`] = `
"/home/josh/script.bas:100:17: warning: identifier has no sigil
  100 print someFn(ident
                   ^"
`;

exports[`given a Inspector > when it is a line > it formats a SyntaxError 1`] = `
"/home/josh/script.bas:100:22: error: expected )
  100 print someFn(ident
                        ^"
`;

exports[`given a Inspector > when it is a line > it formats a SyntaxWarning 1`] = `
"/home/josh/script.bas:100:22: warning: expected )
  100 print someFn(ident
                        ^"
`;

exports[`given a Inspector > when it is not a line > it formats a ParseError 1`] = `
"/home/josh/script.bas:<R0>:22 (after line 100): error: expected )
  print someFn(ident
                        ^
/home/josh/script.bas:<R0>:17 (after line 100): warning: identifier has no sigil
  print someFn(ident
                   ^"
`;

exports[`given a Inspector > when it is not a line > it formats a ParseWarning 1`] = `
"/home/josh/script.bas:<R0>:17 (after line 100): warning: identifier has no sigil
  print someFn(ident
                   ^"
`;

exports[`given a Inspector > when it is not a line > it formats a SyntaxError 1`] = `
"/home/josh/script.bas:<R0>:22 (after line 100): error: expected )
  print someFn(ident
                        ^"
`;

exports[`given a Inspector > when it is not a line > it formats a SyntaxWarning 1`] = `
"/home/josh/script.bas:<R0>:22 (after line 100): warning: expected )
  print someFn(ident
                        ^"
`;
